lib/pbench/test/unit/server/query_apis/test_datasets_contents.py

```diff
@@ -0,0 +1,412 @@
+                            "mtime": "2021-05-01T24:00:00",
+                            "mode": "0o755",
+                            "files": [
+                                {
+                                    "name": "default.csv",
+                                    "mtime": "2021-05-01T24:00:00",
+                                    "size": 122,
+                                    "mode": "0o644",
+                                    "type": "reg",
+                                }
+                            ],
+                            "run_data_parent": "ece030bdgfkjasdkf7435e6a7a6be804",
+                            "authorization": {"owner": "1", "access": "private"},
+                            "@timestamp": "2021-05-01T24:00:00",
+                        },
+                    }
+                ],
+            },
+        }
+        index = self.build_index_from_metadata()
+
+        # get_expected_status() expects to read username and access from the
+        # JSON client payload, however this API acquires that information
+        # from the Dataset. Construct a fake payload corresponding to the
+        # attach_dataset fixture.
+        auth_json = {"user": "drb", "access": "private"}
+        expected_status = self.get_expected_status(
+            auth_json, build_auth_header["header_param"]
+        )
+
+        response = query_api(
+            self.pbench_endpoint,
+            self.elastic_endpoint,
+            self.payload,
+            index,
+            expected_status,
+            json=response_payload,
+            status=HTTPStatus.OK,
+            headers=build_auth_header["header"],
+        )
+        if expected_status == HTTPStatus.OK:
+            res_json = response.json
+            expected_result = {
+                "directories": [],
+                "files": [
+                    {
+                        "name": "default.csv",
+                        "mtime": "2021-05-01T24:00:00",
+                        "size": 122,
+                        "mode": "0o644",
+                        "type": "reg",
+                    }
+                ],
+            }
+            assert expected_result == res_json
+
+    def test_empty_query(
+        self,
+        server_config,
+        query_api,
+        pbench_token,
+        build_auth_header,
+        find_template,
+        provide_metadata,
+    ):
+        """
+        Check the construction of Elasticsearch query URI and filtering of the
+        response body. Note that the mock set up by the attach_dataset fixture
+        matches the dataset name to the dataset's owner.
+        """
+        response_payload = {
+            "took": 55,
+            "timed_out": False,
+            "_shards": {"total": 3, "successful": 3, "skipped": 0, "failed": 0},
+            "hits": {
+                "total": {"value": 0, "relation": "eq"},
+                "max_score": None,
+                "hits": [],
+            },
+        }
```

Comment:
This suggests that the specified directory didn't exist in the TOC at all. I'm not sure we should be returning "success" with no files or subdirectories in that case. `NOT_FOUND` would probably be better.

Reply:
Thanks for pointing this out. Should we return NOT_FOUND if we did not find a sub-directory? what if there are just files present inside it? Should we add NOT_FOUND for directories and display lists?

Also, if there is no directory displayed, it will be an empty directory, so how will it lead to an infinite loop?

We can easily check after extracting sub-directories and files. And if any one of them is not there then we can return NOT_FOUND. What way should we return the `NOT_FOUND`.

Reply:
`NOT_FOUND` if there are no hits in the Elasticsearch response; the specified directory name did not appear as either "parent" (of subdirectories) or "directory" (with files) documents in the selected indices. If there are either files or subdirectories, we want to return success with the two lists, either of which might be empty.

Reply:
> either of which might be empty.

In fact, _both_ of which might be empty:  if we get a hit on the directory which contains no files, then we'll return an empty `files` list; and, if we get no hits on subdirectories; then we'll return an empty `directories` list.  But, in that case (where we did get 1 hit) we should _not_ return `NOT_FOUND`.

We should only return `NOT_FOUND` when there are no hits.

(There is a weird case -- where we find subdirectories without finding the directory itself -- which I _hope_ will never arise...but I expect that your code will handle it properly.)

Reply:
http://pbench.perf.lab.eng.bos.redhat.com/results/ibm-p8-kvm-03-guest-02.virt.pnr.lab.eng.rdu2.redhat.com/pbench-user-benchmark__2022.01.18T06.18.34/1-default/sample1/tools-default/localhost/mpstat/csv/

This might be one of those cases.

Reply:
Right; interesting that the postprocessor created the `csv` directory even though there's nothing in it; but if a client is traversing the directory tree with your new API we definitely want `csv` to show up as a directory in the query for `/default/sample1/tools-default/localhost/mpstat`, and if they ask for `/default/sample1/tools-default/localhost/mpstat/csv` we'd expect `"{directories": [], "files": []}`. The check for `hits` proposed will work fine here; the query should return a single hit, with no files.